### DISCUSSION TITLE
Unable to run python script

### DISCUSSION THREAD

**Discussion by muezz** [vercel/next.js#69830] [Help]
### Summary

I have a python script that I want to run in my nodejs API route. it runs just fine locally but I get the following error in production:
```console
Error: spawnSync /usr/local/bin/python3 ENOENT
    at Object.spawnSync (node:internal/child_process:1124:20)
    at spawnSync (node:child_process:876:24)
    at O (/var/task/.next/server/app/api/webhooks/pins/route.js:1:3170)
    at /var/task/.next/server/app/api/webhooks/pins/route.js:1:2586
    at Array.map (<anonymous>)
    at P (/var/task/.next/server/app/api/webhooks/pins/route.js:1:2575)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async /var/task/node_modules/.pnpm/next@14.2.4_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/next/dist/compiled/next-server/app-route.runtime.prod.js:6:36258
    at async eR.execute (/var/task/node_modules/.pnpm/next@14.2.4_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/next/dist/compiled/next-server/app-route.runtime.prod.js:6:26874)
    at async eR.handle (/var/task/node_modules/.pnpm/next@14.2.4_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/next/dist/compiled/next-server/app-route.runtime.prod.js:6:37512) {
  errno: -2,
  code: 'ENOENT',
  syscall: 'spawnSync /usr/local/bin/python3',
  path: '/usr/local/bin/python3',
  spawnargs: [
    '/var/task/scripts/pins.py',
    'combine_images_with_text'
  ]
}
```

Here is how I run the code:
```typescript
const getMergedPin = async () => {
  const env = Object.assign({}, process.env);
  env.PYTHONPATH = path.join(
    process.cwd(),
    "./scripts/lib/python3.12/site-packages"
  );

  const pythonProcess = spawnSync(
    "/usr/local/bin/python3",
    [
      // path of script file
      path.join(process.cwd(), "./scripts/pins.py"),
      // name of function
      "combine_images_with_text",
    ],
    { env: env, maxBuffer: 10 * 1024 * 1024 }
  );
  console.log(pythonProcess.status);
  console.log(pythonProcess.stderr);
  console.log(pythonProcess.stdout);
  console.log(pythonProcess.error);
  if (pythonProcess.error) {
    throw pythonProcess.error;
  }
  if (pythonProcess.status != 0) {
    throw new Error("Something went wrong");
  }
  const base64String = pythonProcess.stdout.toString().trim();
  if (!base64String) {
    throw new Error("Empty base64 string");
  }
  return { data: base64String };
};
```

I ensure that the required python libraries are installed by modifying the `package.json` file with this build command: `sh setup-scripts.sh && next build`. This is the `setup-scripts.sh` file:
```bash
#!/bin/bash

PYTHON_PATH=$(which python3)
echo "Python3 found at: $PYTHON_PATH"

# Echo the environment variables to check if the path is included
echo "Current PATH environment variable: $PATH"

python3 -m venv ./scripts
source ./scripts/bin/activate
python3 -m pip install --upgrade Pillow
python3 -m pip install --upgrade requests
```

This script does run before deployment and it shows that `python3` is installed at `/usr/local/bin/python3` and the PATH env variables contains `/usr/local/bin`.

### What could be the reason for the script not running?

**Comment by Shriharsh-Deshmukh** [comment]
The issue likely stems from the production environment not having Python3 at the expected location (`/usr/local/bin/python3`). Here are potential solutions:

1. **Check Python Path:** Verify Python3's path in production. Use `which python3` in your setup script and log it to confirm the correct path.

2. **Environment Path Issue:** Ensure the deployment process doesn't override your PATH settings. Explicitly add Python3's path to PATH in your setup-scripts.sh and verify it's active during your Node.js runtime.

3. **Alternative Python Invocation:** Instead of hardcoding `/usr/local/bin/python3`, dynamically determine the Python path at runtime, or ensure Python is globally available via `python3`.

These should help resolve the `ENOENT` error during production.

**Reply by muezz** [reply]
I have tried all 3 approaches already. If you see the `setup-scripts.sh` script that runs before build, you will see that I am using `which python3` to check the path for python. This comes out to be `/usr/local/bin/python3`. I also check the env variables and `/usr/local/bin` was there. I also tried with just `python3` and that did not work.

As another comment suggests, I think its just not possible for some reason. In the end, I gave up after 4 days of trying and moved to `fly.io` where I deployed a FastAPI server that runs the script. 

For anyone that stumbles here in the future, you can either do what I did or deploy an AWS Lambda function as that also allows python.

**Comment by icyJoseph** [comment]
IIRC, python3 is not available in Vercel deployments, might have changed, but here's an older answer, https://github.com/vercel/next.js/discussions/46223#discussioncomment-5073706

To be precise, the Node.js runtime, doesn't have Python. Vercel does support Python deploys, as its own runtime.
